include/deal.II/lac/trilinos_epetra_vector.h

```diff
@@ -0,0 +1,289 @@
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__trilinos_epetra_vector_h
+#define dealii__trilinos_epetra_vector_h
+
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_TRILINOS
+
+#ifdef DEAL_II_WITH_MPI
+
+#include <deal.II/base/index_set.h>
+#include <deal.II/base/std_cxx11/shared_ptr.h>
+#include <deal.II/lac/read_write_vector.h>
+#include <deal.II/lac/vector_space_vector.h>
+#include "mpi.h"
+#include "Epetra_FEVector.h"
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  // Forward declaration
+  template <typename Number>
+  class ReadWriteVector;
+
+  namespace EpetraWrappers
+  {
+    /**
+     * This class implements a wrapper to the the Trilinos distributed vector
+     * class Epetra_FEVector. This class is derived from the
+     * LinearAlgebra::VectorSpaceVector class. Note however that Epetra only
+     * works with Number = double.
+     *
+     * @ingroup TrilinosWrappers
+     * @ingroup Vectors
+     * @author Bruno Turcksin, 2015
+     */
+    class Vector : public VectorSpaceVector<double>
+    {
+    public:
+      /**
+       * Constructor. Create a vector of dimension zero.
+       */
+      Vector();
+
+      /**
+       * Copy constructor. Sets the dimension and the partitioning to that of
+       * the given vector and copies all elements.
+       */
+      Vector(const Vector &V);
+
+      /**
+       * This constructor takes an IndexSet that defines how to distribute the
+       * individual components among the MPI processors. Since it also
+       * includes information abtou the size of the vector, this is all we
+       * need to genereate a %parallel vector.
+       */
+      Vector(const IndexSet &parallel_partitioner,
+             const MPI_Comm &communicator = MPI_COMM_WORLD);
+
+      /**
+       * Reinit functionality. This function destroys the old vector content
+       * and generates a new one based on the input partitioning. The flag
+       * <tt>omit_zeroing_entries</tt> determines whether the vector should be
+       * filled with zero (false) or left untouched (true).
+       */
+      void reinit (const IndexSet &parallel_partitioner,
+                   const MPI_Comm &communicator = MPI_COMM_WORLD,
+                   const bool      omit_zeroing_entries = false);
+
+      /**
+       * Copy function. This function takes a Vector and copies all the
+       * elements. The Vector will have the same parallel distribution as @p
+       * V.
+       */
+      Vector &operator= (const Vector &V);
+
+      /**
+       * Copy function. Copy the elements from the ReadWriteVector @p V in the
+       * Vector. If elements if @p V are not locally owned, there is a
+       * communication.
+       */
+      Vector &operator= (const ::dealii::LinearAlgebra::ReadWriteVector<double> &V);
+
+      /**
+       * Multiply the entire vector by a fixed factor.
+       */
+      virtual VectorSpaceVector<double> &operator*= (const double factor);
```

Comment:
I think you might be able to use covariant return types and return a ```Vector``` here instead. Can you try?